watchtower-plugin/src/dbm.rs

```diff
@@ -209,36 +209,43 @@ impl DBM {
         Some(tower)
     }
 
-    /// Loads the latest registration receipt for a given tower.
-    ///
+    /// Loads the registration receipt(s) for a given tower in the given subscription range.
+    /// If no range is given, then loads the latest receipt
     /// Latests is determined by the one with the `subscription_expiry` further into the future.
     pub fn load_registration_receipt(
         &self,
         tower_id: TowerId,
         user_id: UserId,
-    ) -> Option<RegistrationReceipt> {
-        let mut stmt = self
-            .connection
-            .prepare(
-                "SELECT available_slots, subscription_start, subscription_expiry, signature
-                    FROM registration_receipts 
-                    WHERE tower_id = ?1 AND subscription_expiry = (SELECT MAX(subscription_expiry) 
-                        FROM registration_receipts 
-                        WHERE tower_id = ?1)",
-            )
-            .unwrap();
+        subscription_start: Option<u32>,
+        subscription_expiry: Option<u32>,
+    ) -> Option<Vec<RegistrationReceipt>> {
+        let mut query = "SELECT available_slots, subscription_start, subscription_expiry, signature FROM registration_receipts WHERE tower_id = ?1".to_string();
+
+        let tower_id_encoded = tower_id.to_vec();
+        let mut params: Vec<&dyn ToSql> = vec![&tower_id_encoded];
+
+        if subscription_expiry.is_none() {
+            query.push_str(" AND subscription_expiry = (SELECT MAX(subscription_expiry) FROM registration_receipts WHERE tower_id = ?1)")
+        } else {
+            query.push_str(" AND subscription_start>=?2 AND subscription_expiry <=?3");
+            params.push(&subscription_start);
+            params.push(&subscription_expiry)
+        }
+        let mut stmt = self.connection.prepare(&query).unwrap();
 
-        stmt.query_row([tower_id.to_vec()], |row| {
-            let slots: u32 = row.get(0).unwrap();
-            let start: u32 = row.get(1).unwrap();
-            let expiry: u32 = row.get(2).unwrap();
-            let signature: String = row.get(3).unwrap();
+        stmt.query_map(params.as_slice(), |row| {
+            let slots: u32 = row.get(0)?;
+            let start: u32 = row.get(1)?;
+            let expiry: u32 = row.get(2)?;
+            let signature: String = row.get(3)?;
 
             Ok(RegistrationReceipt::with_signature(
                 user_id, slots, start, expiry, signature,
             ))
         })
-        .ok()
+        .unwrap()
+        .map(|r| r.ok())
+        .collect()
```

Comment:
I can't really get my head around how `collect()` produces `Option<Vec<` here :confused:,
but aside from that, I see that the failure that's being accounted for in `map(|r| r.ok())` is for the de-serialization of the data from the DB (`slots`, `start`, ...) which we assume shouldn't fail anyways.

I think we can `unwrap` while de-serializing directly since this shouldn't actually fail.
and substitute `map(|r| r.ok())` with `map(|r| r.unwrap())`. This will make this method return a `Vec<` instead.

Reply:
I think this was previous approach, but then sr-gi suggested to keep it `Option<Vec<`
 [ref](https://github.com/talaia-labs/rust-teos/pull/199#discussion_r1146028796)

Reply:
I think my point here was being able to distinguish between when no data was found because the tower was not even there, and when the range didn't produce any output. However, we are checking that afterward in `main.rs` by checking if the tower was part of the state. So this may indeed be simplified.

Reply:
Yeah, that is my point, that the suggestion was for that, but we really don't do that here, but always return `Some(...)` and check afterward.

I'm happy with either one or the other.